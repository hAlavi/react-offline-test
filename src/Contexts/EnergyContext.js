import React from 'react';
import PropTypes from 'prop-types';
import axios from '../Helpers/axios';
import config from '../Helpers/config';
import defaultState from './example_api_response.json';

const EnergyContext = React.createContext(defaultState);

class EnergyProvider extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.state = { ...props.energyContext };
  }

  retrieveContext = () => {
    this.setState({
      loading: true,
    });

    axios
      .get(config.energy_service_url)
      .then(response => {
        this.setState({
          loading: false,
          energy: {
            ...response.data.body,
          },
        });
      })
      .catch(error => {
        const errorDetails = {};
        if (error.response) {
          errorDetails.errorCode = error.response.status;
          errorDetails.errorMessage = error.response.data.errorMessage;
        } else {
          errorDetails.errorCode = -1;
          errorDetails.errorMessage = `Request to retrieve energy details failed due to ${error.message}`;
        }

        this.setState({
          loading: false,
          energy: {
            ...errorDetails,
          },
        });
      });
  };

  render() {
    const { children } = this.props;
    return (
      <EnergyContext.Provider
        value={{
          ...this.state,
          retrieveContext: this.retrieveContext,
        }}
      >
        {children}
      </EnergyContext.Provider>
    );
  }
}

EnergyProvider.propTypes = {
  children: PropTypes.node.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  energyContext: PropTypes.object,
};

EnergyProvider.defaultProps = {
  energyContext: defaultState,
};

const EnergyConsumer = EnergyContext.Consumer;

export { EnergyProvider, EnergyConsumer };
