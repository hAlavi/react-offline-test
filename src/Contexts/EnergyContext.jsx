import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import axios from '../Helpers/axios';
import config from '../Helpers/config';
import log from '../Helpers/logger';

const ErrorMessage = styled.p`
  color: red;
  font-size: 24px;
`;

const NS_CONTEXT = 'KiwiPower:EnergyContext';

// Uncomment this block to use sample response as a defaultState
// import defaultResponse from './example_api_response.json';
// const defaultState = {
//   loading: false,
//   energy: defaultResponse,
//   retrieveEnergy: () => {},
// };

// Default State of the context
const defaultState = {
  loading: false,
  failed: false,
  energy: {
    data: {
      generationmix: [
        { fuel: 'testFuel1', perc: 10 },
        { fuel: 'testFuel2', perc: 20 },
      ],
    },
  },
  retrieveEnergy: () => {},
};

const EnergyContext = React.createContext(defaultState);

/**
 * Class EnergyProvider to pass the data to the consumer component
 * @see React.createContext
 */
class EnergyProvider extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.state = { ...props.energy };
  }

  retrieveEnergy = () => {
    this.setState({
      loading: true,
    });

    axios
      .get(config.energy_service_url)
      .then(response => {
        log(NS_CONTEXT, response);
        this.setState({
          loading: false,
          failed: false,
          energy: {
            ...response.data,
          },
        });
      })
      .catch(error => {
        const errorDetails = {};
        //
        if (error.response) {
          errorDetails.errorCode = error.response.status;
          errorDetails.errorMessage = error.response.data.errorMessage;
        } else {
          errorDetails.errorCode = -1;
          errorDetails.errorMessage = `Request to retrieve energy details failed due to ${error.message}`;
        }
        log(NS_CONTEXT, errorDetails.errorMessage);

        this.setState({
          loading: false,
          failed: true,
          energy: {
            ...errorDetails,
          },
        });
      });
  };

  render() {
    const { children } = this.props;
    const { failed } = this.state;

    if (failed) {
      const {
        energy: { errorMessage },
      } = this.state;
      return (
        <ErrorMessage data-testid="error-message">{errorMessage}</ErrorMessage>
      );
    }
    return (
      <EnergyContext.Provider
        value={{
          ...this.state,
          retrieveEnergy: this.retrieveEnergy,
        }}
      >
        {children}
      </EnergyContext.Provider>
    );
  }
}

EnergyProvider.propTypes = {
  children: PropTypes.node.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  energy: PropTypes.object,
};

EnergyProvider.defaultProps = {
  energy: defaultState,
};

const EnergyConsumer = EnergyContext.Consumer;

export { EnergyProvider, EnergyConsumer };
