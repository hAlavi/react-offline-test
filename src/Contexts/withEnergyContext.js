import React from 'react';
import { EnergyConsumer } from './EnergyContext';

const withEnergyContext = Component => {
  const EnergyContextComponent = props => {
    return (
      <EnergyConsumer>
        {energy => <Component energy={energy} {...props} />}
      </EnergyConsumer>
    );
  };

  return EnergyContextComponent;
};

export default withEnergyContext;
