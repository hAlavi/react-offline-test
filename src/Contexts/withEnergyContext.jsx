import React from 'react';
import { EnergyConsumer } from './EnergyContext';

/**
 * Return wrapped component with the energy context
 * @param {React.Component} Component Component to be wrapped
 * @returns {React.Component} wrapped component
 */
const withEnergyContext = Component => {
  const EnergyContextComponent = () => {
    return (
      <EnergyConsumer>
        {/* eslint-disable-next-line react/jsx-props-no-spreading */}
        {energy => <Component energy={energy} />}
      </EnergyConsumer>
    );
  };

  return EnergyContextComponent;
};

export default withEnergyContext;
