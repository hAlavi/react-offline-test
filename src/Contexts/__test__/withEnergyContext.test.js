/* eslint-disable react/prop-types */
import React from 'react';
import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { EnergyProvider } from '../EnergyContext';
import withEnergyContext from '../withEnergyContext';
import 'jest-styled-components';

// Simple Consumer to test the provider/consumer
const SimpleConsumer = ({ energy }) => <>{JSON.stringify(energy)}</>;

afterEach(cleanup);

describe('Unit tests for the withEnergyContext - Tests are similar to EnergyContext tests', () => {
  describe('Visual Snapshot testing (Functional testing have been done in EnergyContext.test.js)', () => {
    it('Renders without crashing', () => {
      const WrappedConsumer = withEnergyContext(SimpleConsumer);
      const { container } = render(
        <EnergyProvider>
          <WrappedConsumer />
        </EnergyProvider>,
      );
      expect(container).toMatchSnapshot();
    });
  });
});
