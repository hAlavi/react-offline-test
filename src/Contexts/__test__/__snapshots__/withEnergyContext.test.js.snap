// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for the withEnergyContext - Tests are similar to EnergyContext tests Visual Snapshot testing (Functional testing have been done in EnergyContext.test.js) Renders without crashing 1`] = `
<div>
  {"loading":false,"failed":false,"energy":{"data":{"generationmix":[{"fuel":"testFuel1","perc":10},{"fuel":"testFuel2","perc":20}]}}}
</div>
`;
